docs/tekton-bundle-contracts.md

```diff
@@ -0,0 +1,38 @@
+<!--
+---
+linkTitle: "Tekton Bundle Contracts"
+weight: 8
+---
+-->
+# Tekton Bundle Contract
+
+When using a Tekton Bundle in a task or pipeline reference, the OCI artifact backing the
+bundle must adhere to the following contract.
+
+## Contract
+
+Only Tekton CRDs (eg, `Task` or `Pipeline`) may reside in a Tekton Bundle used as a Tekton
+bundle reference.
+
+Each layer of the image must map 1:1 with a single Tekton resource.
+
+Each layer must contain the following annotations:
+
+- `org.opencontainers.image.title` => `ObjectMeta.Name` of the resource
+- `cdf.tekton.image.kind` => `TypeMeta.Kind` of the resource, all lowercased (eg, `task`)
+- `cdf.tekton.image.apiVersion` => `TypeMeta.APIVersion` of the resource (eg 
+"tekton.dev/v1alpha1")  
+
+Each { `apiVersion`, `kind`, `title` } must be unique in the image. No resources of the
+same version and kind can be named the same.
+
+The contents of each layer must be the parsed YAML of the corresponding Tekton resource. If
```

Comment:
Is the layer still a gzipped tar file, or just the raw YAML?

Reply:
Good point. This is a bit confusing as worded. The intention is that you should inject a single object into each layer in YAML/JSON form and then perform the standard compression/digesting required by the OCI spec. I will reword this to be more clear

Reply:
I looked at the code to see what actually happens, it might be worth specifying a bit more here. Each layer is a raw yaml file, we don't do any "tar-ing". That then gets compressed as it is stored in the registry.

Reply:
I like this approach - but there are some cons to think through:

- Do we ever think we'll want to have multiple files per bundle? We can't do that without a tar.
- "docker pull/docker save" and "crane pull" expect image layers to be tars. The mime type is actually `application/vnd.oci.image.layer.v1.tar+gzip` - so clients might be expecting a tar here. We could change the mime type, but that might not be supported everywhere.

Reply:
Admittedly my knowledge of the underlying OCI standard is minimal but I will give it a go based on what I think I know. My understanding is that the `go-containerregistry` tool before uploading the image, performs the necessary compression and tar'ing to make the image conform to the standard expected by the registry. If that is the case, I believe that means that while we store and read YAML bytes, transiently there is some compressing/decompressing happening.

Also, I am able to store multiple files in a bundle by adding each one as a new layer with the experimental tool I have built. I think that semi-validates my half-understanding that there is some compression happening in the background but I am happy to be wrong here.

Reply:
So go-containerregistry will compress, but not archive the layer. The mime type it sets by default is tar+gzip, but it's up to you to actually put these bytes into a tar and extract them from a tar when you call "Uncompressed".

I think we have two choices:

- Keep things the way they are (one object per layer, raw yaml bytes without tar-ing). Change the mime/media type of the layer. You can do this with a PR I just sent to go-containerregistry: https://github.com/google/go-containerregistry/commit/071a121b9eeef0f64a5e1238de76ee35d2929082
- Tar files up when you put them in the layer, and keep the mime/media type the same. You can do this with https://golang.org/pkg/archive/tar/#NewReader

cc @jonjohnsonjr, the resident OCI spec expert.

Reply:
This is excellent context that I didn't have so thanks for sharing! Based on your description, to me the latter option makes the most sense because it seems like a waste of space to store uncompressed files in a registry. I will add a note to the contract that the layers must be tar'ed.